### DISCUSSION TITLE
Linux (Desktop) doesn't show Images ?

### DISCUSSION THREAD

**Discussion by pkarsten** [unoplatform/uno#17244] [General]
Hi, 

i created an uno App for Android, Desktop, Windows , I'm devloping under windows with Visual Studio 2022) 

`
![image](https://github.com/unoplatform/uno/assets/47579645/8056c15a-e034-4cbe-b0d1-599c49536236)

`
i try stream an image from OneDrive, but it doesn't show me the image, now i tried as Source an Asset Image, but it does not show me the image, running on the Linux (Raspberry Pi) Machine, in Emulation for Desktop,Android,Win on my windows machine shows it. 

`    <Image  x:Name="BGImage"  Source="{Binding DashImage, Mode=OneWay}" VerticalAlignment="Center" HorizontalAlignment="Center"  />
    <Image x:Name="testImage"  Source="Assets/Images/IMG_0135.JPG" VerticalAlignment="Center" HorizontalAlignment="Center" Width="200" Height="200"/>

    public BitmapImage DashImage    {
        get ; set;
    }
`

what i have forgetten? What can i do? 

Thanks

**Comment by lindexi** [comment]
@pkarsten Could you add more info? Such as all your solution code

**Reply by pkarsten** [reply]
hello lindexi, 

here is my code: 
https://github.com/pkarsten/UnoDesktopPicture

thanks

**Reply by lindexi** [reply]
@pkarsten Do you mean the BGImage in SecondPage.xaml do not show anything?

**Comment by pkarsten** [comment]
Hello Lindexi, 
exactly, it shows neither testImage nor BGImage, testimage is in the Assets directory, BGImage is received via a stream

**Reply by lindexi** [reply]
I can not debug your code, because I fail to download anything.

**Comment by pkarsten** [comment]
hi lindexi, 
i put now my clientid in the git hub code. It works as follow: 
When APp starts, it asked for MSAL Auth , after login, put OneDrive Folder containing images in MyService: 
folder = await graphService.GetPhotosAndImagesFromFolder("/Familienbereich/Bilder");
Click on Mainpage loads the Pictures in a local database, and on second page onClick it will retrieve one pic from database, and stream it from onedrive 

as i say, on my Windows 10 develop maschine, with visual studio 2022 it works fine in all emulations, only on the raspberry Pi it doesn't work, and i doesnt kno how can i debug it remotly on the raspberry pi for see what perhaps happens

**Reply by lindexi** [reply]
@pkarsten Can you sure it can download the image success on the raspberry Pi?

**Comment by pkarsten** [comment]
Lindexi, i dont know If in Raspberry IT receives the Image, but with the other, Testimage, the Image is in assets folder, and this doesnt Show too on the Raspberry.

**Reply by lindexi** [reply]
@pkarsten The cause could be a broken decoder

**Reply by pkarsten** [reply]
lindexi, how can i find out this ? installing linux on raspberry new? 
few days ago, i had the similar app, it was programmed for uwp and skia gtk, and i had the same phaenomen on the raspberry, no images, the i thought is because of the skia gtk, and i develop the app new for desktop , well i will try it install it new  ... thanks for you help

**Reply by lindexi** [reply]
@pkarsten Could you write the console demo to run on raspberry ?

```csharp
SkiaSharp.SKBitmap bmp = SkiaSharp.SKBitmap.Decode(@"TheImageFile.png");
// Check the bmp value 
```

If the bmp be null, it means that the Skia can not decode successfully.

**Reply by pkarsten** [reply]
hi lindexi, 
i get follow error if i try run this project: 
https://github.com/pkarsten/MyUnoConsoleApp



Unhandled exception. System.TypeInitializationException: The type initializer for 'SkiaSharp.SKAbstractManagedStream' threw an exception.
 ---> System.DllNotFoundException: Unable to load shared library 'libSkiaSharp' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable:
/opt/dotnet/shared/Microsoft.NETCore.App/8.0.6/libSkiaSharp.so: cannot open shared object file: No such file or directory
/home/admin/apps/test/libSkiaSharp.so: cannot open shared object file: No such file or directory
/opt/dotnet/shared/Microsoft.NETCore.App/8.0.6/liblibSkiaSharp.so: cannot open shared object file: No such file or directory
/home/admin/apps/test/liblibSkiaSharp.so: cannot open shared object file: No such file or directory
/opt/dotnet/shared/Microsoft.NETCore.App/8.0.6/libSkiaSharp: cannot open shared object file: No such file or directory
/home/admin/apps/test/libSkiaSharp: cannot open shared object file: No such file or directory
/opt/dotnet/shared/Microsoft.NETCore.App/8.0.6/liblibSkiaSharp: cannot open shared object file: No such file or directory
/home/admin/apps/test/liblibSkiaSharp: cannot open shared object file: No such file or directory

   at SkiaSharp.SkiaApi.sk_managedstream_set_procs(SKManagedStreamDelegates procs)
   at SkiaSharp.SKAbstractManagedStream..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKAbstractManagedStream..ctor(Boolean owns)
   at SkiaSharp.SKManagedStream..ctor(Stream managedStream, Boolean disposeManagedStream)
   at SkiaSharp.SKCodec.WrapManagedStream(Stream stream)
   at SkiaSharp.SKCodec.Create(Stream stream, SKCodecResult& result)
   at SkiaSharp.SKCodec.Create(Stream stream)
   at SkiaSharp.SKBitmap.Decode(Stream stream)
   at Program.<Main>$(String[] args) in C:\Users\pkars\source\repos\MyUnoConsoleApp\Program.cs:line 25
Aborted
admin@raspberrypi:~/apps/test $

**Reply by lindexi** [reply]
@pkarsten The exception means that you miss the `libSkiaSharp` dependencies. Could you add `SkiaSharp.NativeAssets.Linux.NoDependencies` nuget library to your project? Or you can install the skia dependencies to your system.

**Reply by jeromelaban** [reply]
_Please make sure to answer in the same thread, it makes the conversation difficult to follow otherwise, thanks!_

> What does that mean now? Do I have to use the SKBitmap as image type for my UI app on the Raspberry Linux? BitmapImage still does not work, nothing is displayed. Do I then have to explicitly handle the other image type for SKIA?

`BitmapImage` is working well on a raspberry. You're most likely having another error, whether it being the image not found on disk, or something else. The repository (https://github.com/pkarsten/UnoDesktopPicture) is not public, so we cannot troubleshoot. 

You do not need to use SKImage or other skia primitives to show images, Uno does it already.

You may also [enable debug logging](https://platform.uno/docs/articles/logging.html) to see more messages around image loading.

**Reply by pkarsten** [reply]
Hi jerome, UnoDesktopPicture is now Public... 
i don't know how can i find out what's is wrong, i also took uno take care of everything... 

i have now a very very simple APP, make it with uno template, and only add one Image to the xaml page: 
https://github.com/pkarsten/TestPic

and this doesn't show me the picture, how can i get this work? or can someone provide me with an example project that I can try out on my raspberry? 
thanks

**Reply by lindexi** [reply]
@pkarsten Yeah, I download your project. And I can repro your problem.

**Reply by lindexi** [reply]
@jeromelaban ~~It seems the UNO's bug~~. The image does not show untill I touch application on my touch screen. 

It can repro in `5.2.161` and `5.2.132` version.

Oh, I know why. It is my OpenGL's issues... It can be fixed by use software render by:

```csharp
// TestPic\Platforms\Desktop\Program.cs 
// Main method
FeatureConfiguration.Rendering.UseOpenGLOnX11 = false;
```

**Reply by lindexi** [reply]
@pkarsten Could you replace `<Image Source="TheImageFile.jpg.jpg" Width="200" Height="220"></Image>` with `<Image Source="ms-appx:///TheImageFile.jpg.jpg" Width="200" Height="220"></Image>` ?

**Reply by lindexi** [reply]
@pkarsten Could you try set `FeatureConfiguration.Rendering.UseOpenGLOnX11 = false;` to use the software render?

**Reply by jeromelaban** [reply]
@pkarsten I build and ran your app, it's working properly for me, on a Pi 3 directly (using the framebuffer renderer):

![image](https://github.com/unoplatform/uno/assets/5839577/ba7f035f-68b1-43d0-b00c-dbc6c96e37a7)

The issue may be related to some build options. Where are you building your app? Directly on the Pi, or on a separate machine? Which command line do you use?

**Reply by pkarsten** [reply]
Hi jerome, i develop on a Windows Machine, with Visual Studio 2022. 

1. I Create an Uno App with Powershell Command: 
_dotnet new unoapp -o TestPic -preset "recommended" -platforms "desktop" -theme-service False -vscode False -pwa False -theme "fluent" -presentation "mvvm" -config False -di False -log "none" -nav "blank" -http False -loc False -tests "none" -toolkit False -dsp False -id "com.companyname.TestPic" -pub "O=TestPic"_

2. In Visual Studio i make my changes

3. with Powershell i Publish: 
dotnet publish -f net8.0-desktop --runtime linux-arm64 --self-contained (tried also no self contained, do not yet know the difference)

4.Copy my files with Powershell to raspberry: 
_scp -r 'C:\Users\pkars\source\repos\testpic\TestPic\bin\Release\net8.0-desktop\linux-arm64\publish\*' admin@192.168.20.111:/home/admin/apps/testpic_

5. On Raspberry i open console and start it with: ./TestPic, No picture is shown :-( 

On Raspberry (3B+)  yesterday i installed the Raspberry , with Raspi Pi Manager , a 64 , Debian Bookworm Desktop,


**what is : framebuffer renderer ?**

**Reply by jeromelaban** [reply]
Thanks. I just repeated your steps, and it's working for me still. 

Let's see what the logs are providing, you can enable them in the `App.xaml.cs` with :

```csharp
builder.SetMinimumLevel(LogLevel.Debug);
...
builder.AddFilter("Microsoft.UI.Xaml.Controls.Image", LogLevel.Debug );
builder.AddFilter("Microsoft.UI.Xaml.Media.Imaging", LogLevel.Debug );
```

**Reply by jeromelaban** [reply]
> what is : framebuffer renderer ?

[Framebuffer is a mode to render apps](https://platform.uno/docs/articles/features/using-linux-framebuffer.html) when there's no window manager.

**Reply by lindexi** [reply]
@pkarsten Can I ask the status? Did you solve the problem?

**Reply by pkarsten** [reply]
i'm trying install it all new from scratch, with other Raspberry Pi 3 B+

**Reply by lindexi** [reply]
I install ubuntu to  Raspberry Pi 3. And I use kwin_x11 on ubuntu. And I find my application can not show any content before I touch the TouchScreen.

**Comment by pkarsten** [comment]
Hi lindexi, 

adding SkiaSharp.NativeAssets.Linux.NoDependencies , and publishing as 

`dotnet publish -f net8.0-desktop --runtime linux-arm64 --self-contained `

give me a result for the App (https://github.com/pkarsten/MyUnoConsoleApp/): 

admin@raspberrypi:~/apps/test $ ./MyUnoConsoleApp
Hello, World!
Image width: 5344
Image height: 3008
Image color type: Bgra8888
Image processed and saved as output.jpg
admin@raspberrypi:~/apps/test $

What does that mean now? Do I have to use the SKBitmap as image type for my UI app on the Raspberry Linux? BitmapImage still does not work, nothing is displayed. Do I then have to explicitly handle the other image type for SKIA?

**Comment by pkarsten** [comment]
this time everything was done on the Raspberry, I tried it on 2 different Raspberry Pis, this time without Windows with Visual Studio 2022 in between, unfortunately no picture is displayed, I don't know what to do, maybe try another Linux? 

From Scratch ... Having a Raspberry Pi 3 B+

• [Getting Started with Uno Platform and the Raspberry Pi](https://platform.uno/docs/articles/guides/raspberry-pi/raspberry-pi-intro.html)
	• Raspberry Pi Imager -> 16 GB SD Card -> Raspberry 3B + -> Install Raspberry Pi OS 32 (A Port of Debian Bookworm with Raspberry Desktop) 
	• SSH -> sudo apt update
	• SSH -> sudo apt full-upgrade
	• SSH -> sudo reboot
	• SSH -> wget -O - https://raw.githubusercontent.com/pjgpetecodes/dotnet8pi/main/install.sh | sudo bash
	• SSH -> sudo reboot
	• SSH -> dortnet --info: 
	.NET SDK:
	 Version:           8.0.302
	 Commit:            ef14e02af8
	 Workload version:  8.0.300-manifests.f6879a9a
	 MSBuild version:   17.10.4+10fbfbf2e
	
	Runtime Environment:
	 OS Name:     raspbian
	 OS Version:  12
	 OS Platform: Linux
	 RID:         linux-arm
	 Base Path:   /opt/dotnet/sdk/8.0.302/
	
	.NET workloads installed:
	There are no installed workloads to display.
	
	Host:
	  Version:      8.0.6
	  Architecture: arm
	  Commit:       3b8b000a0e
	
	.NET SDKs installed:
	  8.0.302 [/opt/dotnet/sdk]
	
	.NET runtimes installed:
	  Microsoft.AspNetCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.AspNetCore.App]
	  Microsoft.NETCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.NETCore.App]
	
	Other architectures found:
	  None
	
	Environment variables:
	  DOTNET_ROOT       [/opt/dotnet]
	
	global.json file:
	  Not found

	• SSH -> dotnet new --install Uno.Templates
Success: Uno.Templates::5.2.41 installed the following templates:
	Template Name       Short Name     Language  Tags                               
	------------------  -------------  --------  ------------------------------------------------------------------------------------------
	Cross-Platform ...  unolib         [C#]      Multi-platform/Uno Platform/Library/Android/iOS/Windows/macOS/Linux/WebAssembly
	Cross-Platform ...  unoapp-uitest  [C#]      Multi-platform/Uno Platform/UITest/Android/iOS/macOS/WebAssembly
	Multi-Platform ...  unoapp-uwp     [C#]      Multi-platform/Uno Platform/Android/iOS/Windows/macOS/Linux/WebAssembly
	Uno Maui Embedd...  unomauilib     [C#]      Multi-platform/Uno Platform/Android/iOS/Mac Catalyst/Windows/WinUI
	Uno Platform App    unoapp         [C#]      Multi-platform/Uno Platform/Android/iOS/macOS/Mac Catalyst/Windows/Linux/WebAssembly/WinUI
	
	• SSH -> dotnet new unoapp -o HelloPi -platforms "desktop" && cd HelloPi
	• SSH -> export DISPLAY=:0
	• SSH -> vim MainPage.xaml
 Edit MainPage.xaml add image Tag: <Ima ge Source="Assets/TheImageFile.jpg.jpg" Width="200" Height="220"></Image>, and copy Image to the Assets Folder: scp -r '"C:\Users\pkars\source\temp\TheImageFile.jpg"' admin@192.168.20.194:/home/admin/apps/HelloPi/HelloPi/HelloPi/Assets
	
dotnet publish -f net8.0-desktop -c Release
./HelloPi

If I put login, 
builder.SetMinimumLevel(LogLevel.Debug);
...
builder.AddFilter("Microsoft.UI.Xaml.Controls.Image", LogLevel.Debug );
builder.AddFilter("Microsoft.UI.Xaml.Media.Imaging", LogLevel.Debug );

where can i see logs? can't :( 

And no Image is Shown :-( 

it is exasperating

**Reply by pkarsten** [reply]
i have now a skia.gtk app running on the raspberry. 
After adding RenderSurfaceType = Software, it shows me the jpg Picture on the xaml Page. 

`	 var host = new GtkHost(() => new App());
            **host.RenderSurfaceType = RenderSurfaceType.Software;**
            host.Run();`

Can i apply something similar if i have the net8-Desktop Target with SkiaHostBuilder?

`        var host = SkiaHostBuilder.Create()
            .App(() => new App())
            .UseX11()
            .UseLinuxFrameBuffer()
            .UseMacOS()
            .UseWindows()
            .Build();
        host.Run();`

unfortunately I can't debug the whole thing directly on the Raspberry, I tried with Visual Studio Code, and the instructions here: https://www.petecodes.co.uk/developing-uwp-apps-for-the-raspberry-pi-with-uno-platform/, I can debug, the routine runs through, no errors, files are transferred to the Raspberry, but it is not executed, but okay that's another topic.

**Reply by lindexi** [reply]
@pkarsten `FeatureConfiguration.Rendering.UseOpenGLOnX11 = false;`

Program.cs:

```csharp
        FeatureConfiguration.Rendering.UseOpenGLOnX11 = false;

        var host = SkiaHostBuilder.Create()
            .App(() =>
            {
                var app = new App();
                app.Launched += (sender, info) =>
                {
                    AppLaunchedInfoTaskCompletionSource.SetResult(info);
                };
                return app;
            })
            .UseX11()
            .UseLinuxFrameBuffer()
            .UseMacOS()
            .UseWindows()
            .Build();

        host.Run();
```

**Reply by pkarsten** [reply]
@lindexi You Made my Day !!! ❤❤❤
Now the app it's running on the Raspberry Showing Images 

thank you guys so much 💪👍🥳

**Reply by lindexi** [reply]
@pkarsten Yeah, see https://github.com/unoplatform/uno/discussions/17244#discussioncomment-9839644

**Reply by pkarsten** [reply]
How could I have overlooked that, a small but fine detail .... 😯😔🤐😐

thank you for your patience

**Reply by jeromelaban** [reply]
This is an interesting information. This would mean that for some reason, parts of the OpenGL support if broken on this device.

**Reply by lindexi** [reply]
@jeromelaban More interesting information is I find my application can not show any content before I touch the TouchScreen. See https://github.com/unoplatform/uno/discussions/17244#discussioncomment-9857800

**Reply by jeromelaban** [reply]
This is not a behavior we could reproduce, having some additional logs will help.

**Reply by lindexi** [reply]
Thank you @jeromelaban . I didn't find anything helpful in the log at the time. The next time I turn on the machine, I will upload the log.
